File: rewrite-core/src/main/java/com/netflix/rewrite/tree/visitor/refactor/op/RemoveImport.java
/*
 * Copyright 2020 the original authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.netflix.rewrite.tree.visitor.refactor.op;

import com.netflix.rewrite.tree.*;
import com.netflix.rewrite.tree.visitor.MethodMatcher;
import com.netflix.rewrite.tree.visitor.refactor.AstTransform;
import com.netflix.rewrite.tree.visitor.refactor.RefactorVisitor;
import lombok.EqualsAndHashCode;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static com.netflix.rewrite.tree.Formatting.format;
import static com.netflix.rewrite.tree.Tr.randomId;
import static java.util.Collections.emptyList;
import static java.util.stream.Collectors.toList;

@EqualsAndHashCode(callSuper = false, onlyExplicitlyIncluded = true)
public class RemoveImport extends RefactorVisitor {
    @EqualsAndHashCode.Include
    private final String clazz;

    private final MethodMatcher methodMatcher;

    private final Type.Class classType;

    private Tr.Import namedImport;
    private Tr.Import starImport;
    private Tr.Import staticStarImport;

    private final Set<String> referencedTypes = new HashSet<>();
    private final Set<Tr.Ident> referencedMethods = new HashSet<>();
    private final List<Tr.Import> staticNamedImports = new ArrayList<>();

    public RemoveImport(String clazz) {
        this.clazz = clazz;
        this.methodMatcher = new MethodMatcher(clazz + " *(..)");
        this.classType = Type.Class.build(clazz);
    }

    @Override
    public String getRuleName() {
        return "core.RemoveImport{classType=" + clazz + "}";
    }

    @Override
    public List<AstTransform> visitImport(Tr.Import impoort) {
        if (impoort.isStatic()) {
            if (impoort.getQualid().getTarget().printTrimmed().equals(clazz)) {
                if ("*".equals(impoort.getQualid().getSimpleName())) {
                    staticStarImport = impoort;
                } else {
                    staticNamedImports.add(impoort);
                }
            }
        } else {
            if (impoort.getQualid().printTrimmed().equals(clazz)) {
                namedImport = impoort;
            } else if ("*".equals(impoort.getQualid().getSimpleName()) && clazz.startsWith(impoort.getQualid().getTarget().printTrimmed())) {
                starImport = impoort;
            }
        }

        return emptyList();
    }

    @Override
    public List<AstTransform> visitTypeName(NameTree name) {
        Type.Class asClass = TypeUtils.asClass(name.getType());
        if (asClass != null && asClass.getPackageName().equals(classType.getPackageName())) {
            referencedTypes.add(asClass.getFullyQualifiedName());
        }
        return super.visitTypeName(name);
    }

    @Override
    public List<AstTransform> visitMethodInvocation(Tr.MethodInvocation method) {
        if (methodMatcher.matches(method) && method.getType() != null &&
                method.getType().getDeclaringType().getFullyQualifiedName().equals(clazz)) {
            referencedMethods.add(method.getName());
        }
        return super.visitMethodInvocation(method);
    }

    @Override
    public List<AstTransform> visitEnd() {
        List<AstTransform> deletes = new ArrayList<>();
        classImportDeletions(deletes);
        staticImportDeletions(deletes);
        return deletes;
    }

    private void classImportDeletions(List<AstTransform> deletes) {
        if (namedImport != null && referencedTypes.stream().noneMatch(t -> t.equals(clazz))) {
            deletes.addAll(delete(namedImport));
        } else if (starImport != null && referencedTypes.isEmpty()) {
            deletes.addAll(delete(starImport));
        } else if (starImport != null && referencedTypes.size() == 1) {
            deletes.addAll(transform(getCursor().enclosingCompilationUnit(), cu -> cu
                    .withImports(cu.getImports().stream()
                            .map(i -> i == starImport ?
                                    new Tr.Import(randomId(), TreeBuilder.buildName(referencedTypes.iterator().next(), format(" ")), false, i.getFormatting()) :
                                    i
                            )
                            .collect(toList())
                    )
            ));
        }
    }

    private void staticImportDeletions(List<AstTransform> deletes) {
        if(staticStarImport != null && referencedMethods.isEmpty()) {
            deletes.addAll(delete(staticStarImport));
        }
        for (Tr.Import staticImport : staticNamedImports) {
            var method = staticImport.getQualid().getSimpleName();
            if(referencedMethods.stream().noneMatch(m -> m.getSimpleName().equals(method))) {
                deletes.addAll(delete(staticImport));
            }
        }
    }

    private List<AstTransform> delete(Tr.Import impoort) {
        return transform(getCursor().enclosingCompilationUnit(), cu -> cu.withImports(cu.getImports().stream()
                .filter(i -> i != impoort)
                .collect(toList())));
    }
}

File: src/main/kotlin/com/netflix/rewrite/refactor/op/RemoveImport.kt
/**
 * Copyright 2016 Netflix, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.netflix.rewrite.refactor.op

import com.netflix.rewrite.ast.*
import com.netflix.rewrite.refactor.RefactorVisitor
import com.netflix.rewrite.search.MethodMatcher
import java.util.*

class RemoveImport(val clazz: String, override val ruleName: String = "remove-import"):
        RefactorVisitor<Tr.CompilationUnit>() {

    val methodMatcher = MethodMatcher("$clazz *(..)")

    var namedImport: Tr.Import? = null
    var starImport: Tr.Import? = null

    val referencedTypes = HashSet<String>()
    val referencedMethods = HashSet<Tr.Ident>()

    val staticNamedImports = ArrayList<Tr.Import>()
    var staticStarImport: Tr.Import? = null

    val classType = Type.Class.build(clazz)

    override fun visitImport(import: Tr.Import): List<AstTransform<Tr.CompilationUnit>> {
        if (import.static) {
            if (import.qualid.target.printTrimmed() == clazz) {
                if (import.qualid.simpleName == "*")
                    staticStarImport = import
                else
                    staticNamedImports.add(import)
            }
        } else {
            if (import.qualid.printTrimmed() == clazz) {
                namedImport = import
            } else if (import.qualid.simpleName == "*" && clazz.startsWith(import.qualid.target.printTrimmed())) {
                starImport = import
            }
        }

        return emptyList()
    }

    override fun visitIdentifier(ident: Tr.Ident): List<AstTransform<Tr.CompilationUnit>> {
        if(ident.type.asClass()?.packageName() == classType.packageName())
            ident.type.asClass()?.let { referencedTypes.add(it.fullyQualifiedName) }
        return emptyList()
    }

    override fun visitMethodInvocation(meth: Tr.MethodInvocation): List<AstTransform<Tr.CompilationUnit>> {
        if(methodMatcher.matches(meth)) {
            if(meth.type?.declaringType?.fullyQualifiedName == clazz)
                referencedMethods.add(meth.name)
        }
        return super.visitMethodInvocation(meth)
    }

    override fun visitEnd(): List<AstTransform<Tr.CompilationUnit>> =
        classImportDeletions() + staticImportDeletions()

    private fun classImportDeletions(): List<AstTransform<Tr.CompilationUnit>> =
        if (namedImport is Tr.Import && referencedTypes.none { it == clazz }) {
            namedImport!!.delete()
        } else if (starImport is Tr.Import && referencedTypes.isEmpty()) {
            starImport!!.delete()
        } else if (starImport is Tr.Import && referencedTypes.size == 1) {
            transform {
                copy(imports = imports.map {
                    if(it == starImport) {
                        val classImportField = TreeBuilder.buildName(referencedTypes.first(), format(" ")) as Tr.FieldAccess
                        Tr.Import(classImportField, false, it.formatting)
                    }
                    else it
                })
            }
        } else emptyList()

    private fun staticImportDeletions(): ArrayList<AstTransform<Tr.CompilationUnit>> {
        val staticImportFixes = ArrayList<AstTransform<Tr.CompilationUnit>>()
        if(staticStarImport is Tr.Import && referencedMethods.isEmpty()) {
            staticImportFixes.addAll(staticStarImport!!.delete())
        }
        staticNamedImports.forEach { staticImport ->
            val method = staticImport.qualid.simpleName
            if(referencedMethods.none { it.simpleName == method })
                staticImportFixes.addAll(staticImport.delete())
        }
        return staticImportFixes
    }

    private fun Tr.Import.delete(): List<AstTransform<Tr.CompilationUnit>> =
        transform { copy(imports = imports - this@delete) }
}